winevtlog: Implement winevtlog plugin

cosmo0920 opened the pull request:
Signed-off-by: Hiroshi Hatake <hatake@calyptia.com>

<!-- Provide summary of changes -->

<!-- Issue number, if available. E.g. "Fixes #31", "Addresses #42, #77" -->
Related to https://github.com/fluent/fluent-bit/issues/4082 and https://github.com/fluent/fluent-bit/issues/3383

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [x] Example configuration file for the change
```ini
[SERVICE]
    Flush           5
    Daemon          yes
    Log_Level       debug

[INPUT]
    Name            winevtlog # not winlog
    Channels        Application,Security,System,Microsoft-Windows-Time-Service/Operational,Microsoft-Windows-SMBClient/Connectivity
    Interval_Sec    1
    read_existing_events yes
    Use_ANSI        yes
    # If the following line is commented out, string_inserts will not inserted into msgpack payload
    # string_inserts no 
    # If the following line is commented out, system part of Windows EventLog is rendered as XML string
    # Render_Event_As_Xml yes
    db              bookmark.sqlite

[OUTPUT]
    Name            stdout
```
- [x] Debug log output from testing the change

#### XML Rendering

```log
[373] winevtlog.0: [1633879283.582482500, {"System"=>"<Event xmlns='http://schemas.microsoft.com/win/2004/08/events/event'><System><Provider Name='docker'/><EventID Qualifiers='0'>1</EventID><Version>0</Version><Level>4</Level><Task>0</Task><Opcode>0</Opcode><Keywords>0x80000000000000</Keywords><TimeCreated SystemTime='2020-12-21T07:43:24.0210725Z'/><EventRecordID>367</EventRecordID><Correlation/><Execution ProcessID='0' ThreadID='0'/><Channel>Application</Channel><Computer>DESKTOP-G457RDR</Computer><Security/></System><EventData><Data>Loading containers: start.</Data></EventData></Event>", "Message"=>"Loading containers: start.", "StringInserts"=>["Loading containers: start."]}]
```

#### Hashmap rendering

```log
[1] winevtlog.0: [1633914155.328866700, {"ProviderName"=>"Microsoft-Windows-Time-Service", "ProviderGuid"=>"{06EDCFEB-0FD0-4E53-ACCA-A6F8BBF81BCB}", "Qualifiers"=>"", "EventID"=>266, "Version"=>0, "Level"=>4, "Task"=>0, "Opcode"=>0, "Keywords"=>"0x4000000000000000", "TimeCreated"=>"2021-10-11 10:02:33 +0900", "EventRecordID"=>1669, "ActivityID"=>"", "RelatedActivityID"=>"", "ProcessID"=>14692, "ThreadID"=>25024, "Channel"=>"Microsoft-Windows-Time-Service/Operational", "Computer"=>"DESKTOP-G457RDR", "UserID"=>"S-1-5-19", "Message"=>"W32time サービスは、タイム ソースの再検出または時刻の再同期 ( あるいはその両方) を行う通知を受信しました。理由コード:2 システム ティック数: 291555000
理由コードの説明:
0 : 管理者からの明示的な時刻再同期要求
1 : このコンピューターの電源状態の変更
2 : ネットワーク インターフェイスまたはネットワーク トポロジに対する変更
3 : W32time 内での、時刻の同期を要する状態の変更
この通知に続くアクションにより、細かい時刻同期の精度に影響が生じる可能性があります。詳細については、https://go.microsoft.com/fwlink/?linkid=845961 を参照してください。", "StringInserts"=>[2, 291555000]}]
```

<!-- Invoke Fluent Bit and Valgrind as: $ valgrind ./bin/fluent-bit <args> -->
- [N/A] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found
Valgrind does not work on Windows.

**Documentation**
<!-- Docs can be edited at https://github.com/fluent/fluent-bit-docs -->
- [x] Documentation required for this feature

<!--  Doc PR (not required but highly recommended) -->
https://github.com/fluent/fluent-bit-docs/pull/625

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

cosmo0920 commented:
No stale.

leonardo-albertovich commented on plugins/in_winevtlog/pack.c:
```diff
@@ -0,0 +1,601 @@
+    return 0;
+}
+
+static int pack_binary(msgpack_packer *mp_pck, PBYTE bin, size_t length)
```
It seems like we have at least two points where we are doing hex encoding, maybe we should consider a making a generic function we can use in those places?

I'm talking about the UTFF-8 RP by Matt that's about to be merged

leonardo-albertovich replied:
It seems like we have at least two points where we are doing hex encoding, maybe we should consider a making a generic function we can use in those places?

I'm talking about the UTFF-8 RP by Matt that's about to be merged

leonardo-albertovich commented on plugins/in_winevtlog/pack.c:
```diff
@@ -0,0 +1,601 @@
+
+static int pack_hex32(msgpack_packer *mp_pck, int32_t hex)
+{
+    CHAR buffer[32];
```
Can we do something about spacing? as in adding it to separate conditional blocks, variable definition from code, etc?

leonardo-albertovich replied:
Can we do something about spacing? as in adding it to separate conditional blocks, variable definition from code, etc?

leonardo-albertovich commented on plugins/in_winevtlog/pack.c:
```diff
@@ -0,0 +1,601 @@
+    if (st != NULL) {
+        SystemTimeToTzSpecificLocalTime(&tzi, st, &st_local);
+
+        struct tm tm = {st_local.wSecond, st_local.wMinute, st_local.wHour, st_local.wDay, st_local.wMonth-1, st_local.wYear-1900, st_local.wDayOfWeek, 0, 0};
```
This line might need to be broken down.

leonardo-albertovich replied:
This line might need to be broken down.

leonardo-albertovich commented on plugins/in_winevtlog/pack.c:
```diff
@@ -0,0 +1,601 @@
+/* -*- Mode: C; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
```
It might be a good idea to take the parts of this file that do basic type packing and put it in a generic file other modules can use so we can minimize code duplication.

leonardo-albertovich replied:
It might be a good idea to take the parts of this file that do basic type packing and put it in a generic file other modules can use so we can minimize code duplication.

leonardo-albertovich commented on plugins/in_winevtlog/winevtlog.c:
```diff
@@ -0,0 +1,786 @@
+
+BOOL cancel_subscription(struct winevtlog_channel *ch)
+{
+    int result = TRUE;
```
Could we get rid of that variable?

leonardo-albertovich replied:
Could we get rid of that variable?

leonardo-albertovich commented on plugins/in_winevtlog/winevtlog.c:
```diff
@@ -0,0 +1,786 @@
+
+    if (!EvtRender(NULL, hEvent, flags, dwBufferSize, wEventXML, &dwBufferUsed, &count))
+    {
+        if (ERROR_INSUFFICIENT_BUFFER == (status = GetLastError()))
```
Could you please move the assignment out of the clause?

leonardo-albertovich replied:
Could you please move the assignment out of the clause?

leonardo-albertovich commented on plugins/in_winevtlog/winevtlog.c:
```diff
@@ -0,0 +1,786 @@
+            /* return buffer size */
+            *event_size = dwBufferSize;
+            wEventXML = (LPWSTR)flb_malloc(dwBufferSize);
+            if (wEventXML)
```
Could you put the curly brackets in the same line of the if / else statements?

leonardo-albertovich replied:
Could you put the curly brackets in the same line of the if / else statements?

leonardo-albertovich commented on plugins/in_winevtlog/winevtlog.c:
```diff
@@ -0,0 +1,786 @@
+
+cleanup:
+
+    if (wEventXML)
```
Could you add curly brackets to the body of this conditional?

cosmo0920 replied:
Fixed in https://github.com/fluent/fluent-bit/pull/4179/commits/a6583c0c04c813dc128074f8223dd90a3777ce78

leonardo-albertovich replied:
Hey, thanks for addressing that, when I mentioned that one instance I meant fixing all of them and the same goes for the rest of the requests.

cosmo0920 replied:
Oh, I got it.
Really fixed in https://github.com/fluent/fluent-bit/pull/4179/commits/6ec139cc2777b4ffecb917dd9c5933e1c3ec0245

leonardo-albertovich commented on plugins/in_winevtlog/winevtlog.c:
```diff
@@ -0,0 +1,786 @@
+                   &dwPropertyCount)) {
+        status = GetLastError();
+
+        if (ERROR_INSUFFICIENT_BUFFER == status) {
```
Could you invert the terms in this comparison?

cosmo0920 replied:
Fixed in https://github.com/fluent/fluent-bit/pull/4179/commits/a6583c0c04c813dc128074f8223dd90a3777ce78

leonardo-albertovich commented on plugins/in_winevtlog/winevtlog.c:
```diff
@@ -0,0 +1,786 @@
+    char *error_message = NULL;
+
+    // Get the size of the buffer
+    if (!EvtFormatMessage(hMetadata, handle, 0, 0, NULL, EvtFormatMessageEvent, dwBufferSize, pBuffer, &dwBufferUsed)) {
```
You might want to break this line up

cosmo0920 replied:
Fixed in https://github.com/fluent/fluent-bit/pull/4179/commits/a6583c0c04c813dc128074f8223dd90a3777ce78

leonardo-albertovich commented on plugins/in_winevtlog/winevtlog.c:
```diff
@@ -0,0 +1,786 @@
+        }
+    }
+
+    wMessage = _wcsdup(pBuffer);
```
Do we want to verify if this is NULL?

cosmo0920 replied:
Not really want to verify. 
In this context, almost errors are caught up before reaching here.
For safer operation, we should do.

leonardo-albertovich commented on plugins/in_winevtlog/winevtlog.c:
```diff
@@ -0,0 +1,786 @@
+
+PWSTR get_description(EVT_HANDLE handle, LANGID langID, unsigned int *message_size)
+{
+#define BUFSIZE 256
```
Could you rename this with a more representative name and move it to the top of the file?

cosmo0920 replied:
Fixed in https://github.com/fluent/fluent-bit/pull/4179/commits/a6583c0c04c813dc128074f8223dd90a3777ce78

leonardo-albertovich commented on plugins/in_winevtlog/winevtlog.c:
```diff
@@ -0,0 +1,786 @@
+
+    wMessage = get_message(hMetadata, handle, message_size);
+
+#undef BUFSIZE
```
Why do we want to undefine this?

cosmo0920 replied:
Fixed in https://github.com/fluent/fluent-bit/pull/4179/commits/a6583c0c04c813dc128074f8223dd90a3777ce78

leonardo-albertovich commented on plugins/in_winevtlog/winevtlog.c:
```diff
@@ -0,0 +1,786 @@
+        ch->count = count;
+        for (int i = 0; i < count; i++) {
+            ch->events[i] = hEvents[i];
+            EvtUpdateBookmark(ch->bookmark, ch->events[i]);
```
Should we care about this result?

cosmo0920 replied:
Honestly speaking, I'm not sure.
But, carefully handling them would cause an issue to stop the latter processes for obtaining the next events.

leonardo-albertovich commented on plugins/in_winevtlog/winevtlog.c:
```diff
@@ -0,0 +1,786 @@
+        return FLB_FALSE;
+    }
+
+    if (!EvtNext(ch->subscription, SUBSCRIBE_ARRAY_SIZE, hEvents, INFINITE, 0, &count)) {
```
Could you move this so the result is saved to a variable which is used in the conditional?

cosmo0920 replied:
Fixed in https://github.com/fluent/fluent-bit/pull/4179/commits/a02c29553e0fffc9eeb847eecd10a21f3767f9e6

leonardo-albertovich commented on plugins/in_winevtlog/winevtlog.c:
```diff
@@ -0,0 +1,786 @@
+{
+    DWORD status = ERROR_SUCCESS;
+    PWSTR wSystem = NULL;
+    unsigned int system_size = 0, message_size = 0, string_inserts_size = 0;
```
Could you define variables in individual lines?

cosmo0920 replied:
Fixed in https://github.com/fluent/fluent-bit/pull/4179/commits/a6583c0c04c813dc128074f8223dd90a3777ce78

leonardo-albertovich replied:
Did you make this correction in all the occurrences?

cosmo0920 replied:
Yes, I did. Other part does not have multiple variable definitions at the same line, as far as I know.

cosmo0920 replied:
They should be fixed in https://github.com/fluent/fluent-bit/pull/4179/commits/593679caeb13dac5e9e4a52e4c40aa5f133a72a5

leonardo-albertovich commented on plugins/in_winevtlog/winevtlog.c:
```diff
@@ -0,0 +1,786 @@
+    unsigned int read_size = 0;
+    PWSTR wMessage = NULL;
+    PEVT_VARIANT pSystem = NULL;
+    PEVT_VARIANT pStringInserts = NULL;
```
Could we not mix camel case and underscores at least in the same file? I'm against camel case and any windows convention because it makes it harder to get acclimated to the source code when switching between files.

cosmo0920 replied:
Fixed in https://github.com/fluent/fluent-bit/pull/4179/commits/aae73fa0a70fa377af361f379601de7a0fc3a1e7.

leonardo-albertovich commented on plugins/in_winevtlog/winevtlog.c:
```diff
@@ -0,0 +1,786 @@
+        }
+
+        /* Closes any events in case an error occurred above. */
+        for (DWORD i = 0; i < ch->count; i++) {
```
Could you move the definition of this variable to the begining of the function and use a representative name for it?

cosmo0920 replied:
Fixed in https://github.com/fluent/fluent-bit/pull/4179/commits/a6583c0c04c813dc128074f8223dd90a3777ce78

leonardo-albertovich commented on plugins/in_winevtlog/in_winevtlog.c:
```diff
@@ -0,0 +1,257 @@
+
+    /* Initialize context */
+    ctx = flb_calloc(1, sizeof(struct winevtlog_config));
+    if (!ctx) {
```
Could you change this to explicitly compare against NULL? (same goes for other occurrences but I won't spam the diffs with this)

cosmo0920 replied:
Fixed in https://github.com/fluent/fluent-bit/pull/4179/commits/6f9c2a4ee4954743f566183ac7000d995428ec55.

leonardo-albertovich commented on plugins/in_winevtlog/winevtlog.c:
```diff
@@ -0,0 +1,786 @@
+        return -1;
+    }
+
+    snprintf(query, sizeof(query) - 1, SQL_UPDATE_CHANNEL,
```
Should we escape these strings to prevent uninentional sql injection from breaking the query and causing something to be missed? Is that a possibility with the type of data that rnder_event generates?

cosmo0920 replied:
XML encoded channel bookmark information is generated by Windows Event Tracing system.
Also, I'm not sure what should we do to protect suspicious XML encoded bookmarks....

leonardo-albertovich replied:
That's why I mentioned it unintentionally breaking quotation marks, because the character might just be a legitimate part of the data returned.

cosmo0920 replied:
The actual stored channel bookmarks and metadata in the sqlite DB as follows:

```console
sqlite> select * from in_winevtlog_channels;
Application|<BookmarkList>
  <Bookmark Channel='Application' RecordId='33640' IsCurrent='true'/>
</BookmarkList>|1637333345|1633878463
Security|<BookmarkList>
  <Bookmark Channel='Security' RecordId='497867' IsCurrent='true'/>
</BookmarkList>|1637333345|1633878464
System|<BookmarkList>
  <Bookmark Channel='System' RecordId='32359' IsCurrent='true'/>
</BookmarkList>|1637333349|1633878464
Microsoft-Windows-Time-Service/Operational|<BookmarkList>
  <Bookmark Channel='Microsoft-Windows-Time-Service/Operational' RecordId='1800' IsCurrent='true'/>
</BookmarkList>|1637297649|1633878464
Microsoft-Windows-SMBClient/Connectivity|<BookmarkList>
  <Bookmark Channel='Microsoft-Windows-SmbClient/Connectivity' RecordId='23396' IsCurrent='true'/>
</BookmarkList>|1637333345|1633878465
```

The following structure is fixed:

```xml
<BookmarkList>
  <Bookmark Channel='[CHANNEL NAME]' RecordId='[RECORD NUMBER]' IsCurrent='true'/>
</BookmarkList>
```

leonardo-albertovich replied:
And that works because it's predictable data and because of that for that specific field you used double quotation marks as delimiter which is totally fine.

leonardo-albertovich reviewed (changes requested):
I tried to mention each type of issue once to not be too pedantic.
I think the coding style issues are the show real stoppers here.
I didn't even try to ensure that the code is bug free, it's a large patch and I'm not a machine so all I can talk about are the styling choices.

It'd be good if we could prioritize this to resolve this in the near future.

Thank you very much!

cosmo0920 commented:
I almost solved addressed issues except for the followings:

* Return values check for EvtUpdateBookmark: https://github.com/fluent/fluent-bit/pull/4179#discussion_r752289146
* hex encoding: https://github.com/fluent/fluent-bit/pull/4179#discussion_r752295762
* NULL check necessity on get_message `_wcsdup`: https://github.com/fluent/fluent-bit/pull/4179#discussion_r752284006

leonardo-albertovich commented on plugins/in_winevtlog/winevtlog.c:
```diff
@@ -0,0 +1,792 @@
+    DWORD i = 0;
+
+    while (winevtlog_next(ch, hit_threshold)) {
+        for (DWORD i = 0; i < ch->count; i++) {
```
I think you forgot to remove this definition.

cosmo0920 replied:
Fixed in https://github.com/fluent/fluent-bit/pull/4179/commits/f18098bc507d377ad66b547db1e984929fd66a75.

cosmo0920 replied:
Yeah. Actually, I forgot to remove variable definition here. Thanks for the pointing out.

leonardo-albertovich commented on plugins/in_winevtlog/winevtlog.c:
```diff
@@ -0,0 +1,792 @@
+        EvtClose(ch->bookmark);
+        ch->bookmark = NULL;
+    }
+    for (int i = 0; i < ch->count; i++) {
```
Please move this definition to the top

cosmo0920 replied:
Fixed in https://github.com/fluent/fluent-bit/pull/4179/commits/f18098bc507d377ad66b547db1e984929fd66a75.

leonardo-albertovich commented on plugins/in_winevtlog/winevtlog.c:
```diff
@@ -0,0 +1,792 @@
+
+    if (status == ERROR_SUCCESS) {
+        ch->count = count;
+        for (int i = 0; i < count; i++) {
```
Please move this definition to the top

cosmo0920 replied:
Fixed in https://github.com/fluent/fluent-bit/pull/4179/commits/f18098bc507d377ad66b547db1e984929fd66a75.

sean-scott-lr commented:
This PR is valuable Thank you for this work.

question.  Will the calls to EvtRender that you have made in this PR map the encoded values too?

For example: 
Level
Task
Opcode
Keywords
?

cosmo0920 commented:
> question. Will the calls to EvtRender that you have made in this PR map the encoded values too?

EvtRender just retrieve PEVT_VARIANT variables.
If Event format is specified, we can obtain the information which includes EvtRender returned variable like this:
https://github.com/fluent/fluent-bit/pull/4179/files#diff-1881e310fb54efcc6983a5c3a63d357e0d89fe8ad9aa8585b2c1777743e9685cR442-R600

sean-scott-lr commented:
I would like to use fluentbit provided that this PR is merged. 
1) What is the likelihood that will happen
2) Assuming it is merged, how long does it usually take until it makes its way into a released version?

gillg commented:
What would be the difficulty to implement a remote event read with this new API ?
Use it to read logs from inside a windows Docker container would be super helpful.

cosmo0920 commented:
> What would be the difficulty to implement a remote event read with this new API ?
> Use it to read logs from inside a windows Docker container would be super helpful.

With current Fluent Bit, there is a bit of difficulties for implementing remote accessing Windows EventLog feature.
In Fluentd world, we'd already implemented this feature on fluent-plugin-windows-evenlog plugin:
https://github.com/fluent/fluent-plugin-windows-eventlog

gillg commented:
Great to know ! So I suppose there is no compatibility for plugins between fluent[d,bit].
I think I will consider fluentd as local collector...

cosmo0920 commented:
Thanks!
